File: Frontend/mi-app/src/features/onboarding/lib/weeklyAvailability.ts
export type TimeSlot = "mañana" | "tarde" | "noche";

export type WeekDay =
  | "lunes"
  | "martes"
  | "miercoles"
  | "jueves"
  | "viernes"
  | "sabado";

export type WeeklyAvailabilityPayload = {
  "disponibilidad a la semana": Record<WeekDay, TimeSlot[]>;
};

export const WEEK_DAYS: { id: WeekDay; label: string }[] = [
  { id: "lunes", label: "Lunes" },
  { id: "martes", label: "Martes" },
  { id: "miercoles", label: "Miércoles" },
  { id: "jueves", label: "Jueves" },
  { id: "viernes", label: "Viernes" },
  { id: "sabado", label: "Sábado" },
];

export const TIME_SLOTS: { id: TimeSlot; label: string; hours: string }[] = [
  { id: "mañana", label: "Mañana", hours: "7–12" },
  { id: "tarde", label: "Tarde", hours: "1–5" },
  { id: "noche", label: "Noche", hours: "6–10" },
];

export function createEmptyWeeklyAvailability(): WeeklyAvailabilityPayload {
  return {
    "disponibilidad a la semana": {
      lunes: [],
      martes: [],
      miercoles: [],
      jueves: [],
      viernes: [],
      sabado: [],
    },
  };
}

export function countSelectedSlots(payload: WeeklyAvailabilityPayload): number {
  return Object.values(payload["disponibilidad a la semana"]).reduce(
    (total, slots) => total + slots.length,
    0,
  );
}

export function countDaysWithSelection(payload: WeeklyAvailabilityPayload): number {
  return Object.values(payload["disponibilidad a la semana"]).filter(
    (slots) => slots.length > 0,
  ).length;
}

export function getInitiallySelectedDays(payload: WeeklyAvailabilityPayload): WeekDay[] {
  const schedule = payload["disponibilidad a la semana"];
  return WEEK_DAYS.filter(({ id }) => schedule[id].length > 0).map(({ id }) => id);
}

export function clearDaySlots(
  payload: WeeklyAvailabilityPayload,
  day: WeekDay,
): WeeklyAvailabilityPayload {
  return {
    "disponibilidad a la semana": {
      ...payload["disponibilidad a la semana"],
      [day]: [],
    },
  };
}

File: Backend/app/schemas/weekly_availability.py
WEEKLY_AVAILABILITY_KEY = "disponibilidad a la semana"
ALLOWED_WEEK_DAYS = frozenset({"lunes", "martes", "miercoles", "jueves", "viernes", "sabado"})
ALLOWED_TIME_SLOTS = frozenset({"mañana", "tarde", "noche"})


def validate_weekly_availability(value: dict) -> dict:
    if WEEKLY_AVAILABILITY_KEY not in value:
        raise ValueError(f"Falta la clave '{WEEKLY_AVAILABILITY_KEY}'")

    schedule = value[WEEKLY_AVAILABILITY_KEY]
    if not isinstance(schedule, dict):
        raise ValueError(f"'{WEEKLY_AVAILABILITY_KEY}' debe ser un objeto")

    unknown_days = set(schedule.keys()) - ALLOWED_WEEK_DAYS
    if unknown_days:
        raise ValueError(f"Días no permitidos: {', '.join(sorted(unknown_days))}")

    normalized_schedule: dict[str, list[str]] = {}
    for day in ALLOWED_WEEK_DAYS:
        slots = schedule.get(day, [])
        if slots is None:
            slots = []
        if not isinstance(slots, list):
            raise ValueError(f"Los horarios de '{day}' deben ser una lista")

        normalized_slots: list[str] = []
        seen: set[str] = set()
        for slot in slots:
            if not isinstance(slot, str):
                raise ValueError(f"Horario inválido en '{day}'")
            if slot not in ALLOWED_TIME_SLOTS:
                raise ValueError(f"Horario no permitido en '{day}': {slot}")
            if slot in seen:
                continue
            seen.add(slot)
            normalized_slots.append(slot)

        normalized_schedule[day] = normalized_slots

    return {WEEKLY_AVAILABILITY_KEY: normalized_schedule}
